File: server_integration_tests/clients/java_client/client-app/src/test/java/dev/graphlink/server/DefaultsTest.java
package dev.graphlink.server;

import dev.graphlink.server.generated.client.GraphLinkClient;
import dev.graphlink.server.generated.enums.Priority;
import dev.graphlink.server.generated.inputs.ConfigInput;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

public class DefaultsTest {

    private GraphLinkClient client;

    @BeforeEach
    void setUp() {
        client = Fixtures.newClient();
    }

    @Test
    void inputFieldDefaultsAppliedWhenOmitted() {
        var c = client.queries.resolveConfig(ConfigInput.builder().build()).getResolveConfig();
        assertEquals(25, c.getPageSize());
        assertEquals(1.5, c.getRatio());
        assertEquals("asc", c.getSort());
        assertFalse(c.getVerbose());
        assertEquals(Priority.MEDIUM, c.getPriority());
        assertEquals(List.of("default", "seed"), c.getTags());
        assertTrue(c.getEmpties().isEmpty());
        assertNull(c.getNote());
    }

    @Test
    void explicitValuesOverrideDefaults() {
        var c = client.queries.resolveConfig(ConfigInput.builder()
                .pageSize(5).sort("desc").priority(Priority.HIGH).note("x").build()).getResolveConfig();
        assertEquals(5, c.getPageSize());
        assertEquals("desc", c.getSort());
        assertEquals(Priority.HIGH, c.getPriority());
        assertEquals("x", c.getNote());
    }

    @Test
    void nestedObjectArgDefaultMergesWithFieldDefaults() {
        var range = client.queries.resolveRange(null).getResolveRange();
        assertEquals(5, range.getMin());
        assertEquals(100, range.getMax());
    }

    @Test
    void argumentLevelScalarDefaults() {
        assertEquals("Hi world!", client.queries.greet(null, null).getGreet());
        assertEquals("Hi Bob!Hi Bob!", client.queries.greet("Bob", 2).getGreet());
    }

    @Test
    void enumArgumentPassedValueRoundTripsOmittedUsesDefaultHigh() {
        assertEquals(Priority.LOW, client.queries.echoPriority(Priority.LOW).getEchoPriority());
        assertEquals(Priority.HIGH, client.queries.echoPriority(null).getEchoPriority());
    }
}

File: server_integration_tests/clients/kotlin_client/client-app/src/test/kotlin/dev/graphlink/server/DefaultsTest.kt
package dev.graphlink.server

import dev.graphlink.server.generated.client.GraphLinkClient
import dev.graphlink.server.generated.enums.Priority
import dev.graphlink.server.generated.inputs.ConfigInput
import kotlinx.coroutines.test.runTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class DefaultsTest {

    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        client = Fixtures.newClient()
    }

    @Test
    fun inputFieldDefaultsAppliedWhenOmitted() = runTest {
        val c = client.queries.resolveConfig(ConfigInput()).resolveConfig
        assertEquals(25, c.pageSize)
        assertEquals(1.5, c.ratio)
        assertEquals("asc", c.sort)
        assertFalse(c.verbose!!)
        assertEquals(Priority.MEDIUM, c.priority)
        assertEquals(listOf("default", "seed"), c.tags)
        assertTrue(c.empties!!.isEmpty())
        assertNull(c.note)
    }

    @Test
    fun explicitValuesOverrideDefaults() = runTest {
        val c = client.queries.resolveConfig(
            ConfigInput(pageSize = 5, sort = "desc", priority = Priority.HIGH, note = "x"),
        ).resolveConfig
        assertEquals(5, c.pageSize)
        assertEquals("desc", c.sort)
        assertEquals(Priority.HIGH, c.priority)
        assertEquals("x", c.note)
    }

    @Test
    fun nestedObjectArgDefaultMergesWithFieldDefaults() = runTest {
        val range = client.queries.resolveRange().resolveRange
        assertEquals(5, range.min)
        assertEquals(100, range.max)
    }

    @Test
    fun argumentLevelScalarDefaults() = runTest {
        assertEquals("Hi world!", client.queries.greet().greet)
        assertEquals("Hi Bob!Hi Bob!", client.queries.greet(name = "Bob", times = 2).greet)
    }

    @Test
    fun enumArgumentPassedValueRoundTripsOmittedUsesDefaultHigh() = runTest {
        assertEquals(Priority.LOW, client.queries.echoPriority(Priority.LOW).echoPriority)
        assertEquals(Priority.HIGH, client.queries.echoPriority().echoPriority)
    }
}
